mokapot/confidence_writer.py

```diff
@@ -0,0 +1,106 @@
+
+from .tabular_data import TabularDataWriter, SqliteWriter
+
+
+@typechecked
+class ConfidenceSqliteWriter(SqliteWriter):
+    def __init__(
+        self,
+        database: str | Path | sqlite3.Connection,
+        columns: list[str],
+        column_types: list | None = None,
+        level: str = "psms",
+        qvalue_column: str = "q_value",
+        pep_column: str="posterior_error_prob",
+    ) -> None:
+        super().__init__(database, columns, column_types)
+        self.level_cols = {
+            "precursors": ["PRECURSOR_VALIDATION", "PCM_ID", "Precursor"],
+            "modifiedpeptides": [
+                "MODIFIED_PEPTIDE_VALIDATION",
+                "MODIFIED_PEPTIDE_ID",
+                "ModifiedPeptide",
+            ],
+            "peptides": ["PEPTIDE_VALIDATION", "PEPTIDE_ID", "peptide"],
+            "peptidegroups": [
+                "PEPTIDE_GROUP_VALIDATION",
+                "PEPTIDE_GROUP_ID",
+                "PeptideGroup",
+            ],
+        }
+        self.level = level
+        self.qvalue_column = qvalue_column
+        self.pep_column = pep_column
+
+    def get_query(self, level, qvalue_column, pep_column):
+        if level == "psms":
+            query = f"UPDATE CANDIDATE SET PSM_FDR = :{qvalue_column}, SVM_SCORE = :score, POSTERIOR_ERROR_PROBABILITY = :{pep_column} WHERE CANDIDATE_ID = :PSMId;"
+        else:
+            table_name, table_id_col, mokapot_id_col = self.level_cols[level]
+            query = f"INSERT INTO {table_name}({table_id_col},FDR,PEP,SVM_SCORE) VALUES(:{mokapot_id_col},:{qvalue_column},:{pep_column},:score)"
+        return query
+
+    def append_data(self, data):
+        query = self.get_query(self.level, self.qvalue_column, self.pep_column)
+        # todo: what about using connection.executemany()? Should be faster...
+        data = data.to_dict("records")
+        for row in data:
+            self.connection.execute(query, row)
+
+
+@typechecked
+def write_confidences(
+    data_iterator: Iterator[pd.DataFrame],
+    q_value_iterator: list[np.array],
+    pep_iterator: list[np.array],
+    target_iterator: list[np.array],
+    out_paths: list[Path],
+    decoys: bool,
+    level: str,
+    out_columns: list[str],
+    qvalue_column: str="q_value",
+    pep_column: str="posterior_error_prob",
+):
+    if not decoys and len(out_paths) > 1:
+        out_paths.pop(1)
+    is_sqlite = True if out_paths[0].suffix == ".db" else False
+
+    # Create the writers
+    if is_sqlite:
+        create_writer = lambda path: ConfidenceSqliteWriter(path, out_columns, level=level, qvalue_column=qvalue_column, pep_column=pep_column)
+    else:
+        create_writer = lambda path: TabularDataWriter.from_suffix(path, out_columns)
+    writers = [create_writer(path) for path in out_paths]
+    # for writer in writers:
+    #     writer.initialize()
+
+    # Now write the confidence data
+    for data_chunk, qvals_chunk, peps_chunk, targets_chunk in zip(
```

Comment:
This one is a rabbit whole... unintuitively, what is called "data_iterator", and was provided by "get_chunked_data_iterator" is in fact not an iterator but a generator (definition in tabular_data.py). In contrast to an actual iterator, we cannot check the length (`len(list(generator))` would iterate the whole dataset which we want to avoid). Following that:
- we need to work on our naming
- Python 3.10 would be super helpful because would could keep using zip to lazyly evaluate this and still have guarantees that the lists are same length. Would you mind bumping the minimum required Python version to 3.10?

Reply:
Do you really think this is an issue? 
Here only chunked iterators are used and the chunks of the later iterators are combined (joined horizontally) to the first chunk, or used for indexing. So, whenever one of the later chunks (qvals_chunk, peps_chunk or targets_chunk) has a length different than the length of the data_chunk, an exception will be raised. 
So, in almost all cases this will not go unnoticed.
If I see that correctly, an error can only go unnoticed if the length of the first iterator is an integer multiple of the chunk size and one or more of the other iterators are longer than that.

Reply:
I fully agree with your statement about making assumptions explicit. Your `strictzip` function also looks nice.

However, I grepped the code and there are quite a few places where zip is used of which some could also in principle cause problems. What do you think about making an issue out of this and checking all the places then? We could put your function into `util.py` then or use the reference implementation from PEP 618 (https://peps.python.org/pep-0618/).

Reply:
https://github.com/wfondrie/mokapot/pull/119#issuecomment-2231943971 <- Added there to the series of things that should be done but not now